[WebGPU] WebGPU spec updated to support SharedWorkers and ServiceWorkers from only DedicatedWorkers

mwyrzykowski opened the pull request:
#### da1878500441aeae9a3c5683e883f4531b7ec132
<pre>
[WebGPU] WebGPU spec updated to support SharedWorkers and ServiceWorkers from only DedicatedWorkers
<a href="https://bugs.webkit.org/show_bug.cgi?id=273195">https://bugs.webkit.org/show_bug.cgi?id=273195</a>
&lt;radar://126990546&gt;

Reviewed by Tadeu Zagallo.

Add support for SharedWorkers and ServiceWorkers as the spec
requires it now.

* Source/WebCore/Modules/WebGPU/GPU.idl:
* Source/WebCore/Modules/WebGPU/GPUAdapter.idl:
* Source/WebCore/Modules/WebGPU/GPUAdapterInfo.idl:
* Source/WebCore/Modules/WebGPU/GPUBindGroup.idl:
* Source/WebCore/Modules/WebGPU/GPUBindGroupLayout.idl:
* Source/WebCore/Modules/WebGPU/GPUBuffer.idl:
* Source/WebCore/Modules/WebGPU/GPUBufferUsage.idl:
* Source/WebCore/Modules/WebGPU/GPUColorWrite.idl:
* Source/WebCore/Modules/WebGPU/GPUCommandBuffer.idl:
* Source/WebCore/Modules/WebGPU/GPUCommandEncoder.idl:
* Source/WebCore/Modules/WebGPU/GPUCompilationInfo.idl:
* Source/WebCore/Modules/WebGPU/GPUCompilationMessage.idl:
* Source/WebCore/Modules/WebGPU/GPUComputePassEncoder.idl:
* Source/WebCore/Modules/WebGPU/GPUComputePipeline.idl:
* Source/WebCore/Modules/WebGPU/GPUDeviceLostInfo.idl:
* Source/WebCore/Modules/WebGPU/GPUExternalTexture.idl:
* Source/WebCore/Modules/WebGPU/GPUInternalError.idl:
* Source/WebCore/Modules/WebGPU/GPUMapMode.idl:
* Source/WebCore/Modules/WebGPU/GPUOutOfMemoryError.idl:
* Source/WebCore/Modules/WebGPU/GPUPipelineLayout.idl:
* Source/WebCore/Modules/WebGPU/GPUQuerySet.idl:
* Source/WebCore/Modules/WebGPU/GPUQueue.idl:
* Source/WebCore/Modules/WebGPU/GPURenderBundle.idl:
* Source/WebCore/Modules/WebGPU/GPURenderBundleEncoder.idl:
* Source/WebCore/Modules/WebGPU/GPURenderPassEncoder.idl:
* Source/WebCore/Modules/WebGPU/GPURenderPipeline.idl:
* Source/WebCore/Modules/WebGPU/GPUSampler.idl:
* Source/WebCore/Modules/WebGPU/GPUShaderModule.idl:
* Source/WebCore/Modules/WebGPU/GPUShaderStage.idl:
* Source/WebCore/Modules/WebGPU/GPUSupportedFeatures.idl:
* Source/WebCore/Modules/WebGPU/GPUSupportedLimits.idl:
* Source/WebCore/Modules/WebGPU/GPUTexture.idl:
* Source/WebCore/Modules/WebGPU/GPUTextureUsage.idl:
* Source/WebCore/Modules/WebGPU/GPUTextureView.idl:
* Source/WebCore/Modules/WebGPU/GPUUncapturedErrorEvent.idl:
* Source/WebCore/Modules/WebGPU/GPUValidationError.idl:
* Source/WebCore/Modules/WebGPU/WGSLLanguageFeatures.idl:
* Source/WebCore/html/canvas/GPUCanvasContext.idl:
Update idl files

* Source/WebCore/workers/WorkerGlobalScope.h:
* Source/WebCore/workers/WorkerLoaderProxy.h:
* Source/WebCore/workers/WorkerMessagingProxy.h:
* Source/WebCore/workers/WorkerThread.h:
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.cpp:
(WebCore::ServiceWorkerGlobalScope::create):
(WebCore::ServiceWorkerGlobalScope::ServiceWorkerGlobalScope):
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.h:
* Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
(WebCore::ServiceWorkerThread::createWorkerGlobalScope):
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.h:
* Source/WebCore/workers/shared/context/SharedWorkerThreadProxy.h:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::installServiceWorker):
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.h:
* Source/WebKit/WebProcess/Storage/WebSharedWorkerContextManagerConnection.cpp:
(WebKit::WebSharedWorkerContextManagerConnection::launchSharedWorker):
* Source/WebKit/WebProcess/Storage/WebSharedWorkerContextManagerConnection.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createWorkerClient):
* Source/WebKit/WebProcess/WebCoreSupport/WebWorkerClient.cpp:
(WebKit::WebWorkerClient::create):
* Source/WebKit/WebProcess/WebCoreSupport/WebWorkerClient.h:

Canonical link: <a href="https://commits.webkit.org/279984@main">https://commits.webkit.org/279984@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/be06cc92b4c4f2f4ed9a1c0ddea8a3738e1fb6cb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55062 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34527 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7674 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58344 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5794 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57362 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42087 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5831 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44581 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3958 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57157 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32601 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47709 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25705 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29385 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5053 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3938 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/51244 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5288 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59934 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30341 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5440 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52011 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31465 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47777 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51495 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12261 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32491 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31255 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebCore/page/WorkerNavigator.cpp:
```diff
@@ -62,11 +62,16 @@ GPU* WorkerNavigator::gpu()
-                return nullptr;
-
-            RefPtr gpu = workerGlobalScope.graphicsClient()->createGPUForWebGPU();
+            if (auto type = workerGlobalScope.type(); type == WorkerGlobalScope::Type::SharedWorker || type == WorkerGlobalScope::Type::ServiceWorker)
```
The constructor for `SharedWorkerThreadProxy` creates a `WorkerClient` (implementing `GraphicsClient`), so I would expect the `else` path to 'just work' for that case.

I think we could also do the same in `ServiceWorkerThreadProxy` and hopefully not need this first path at all?

mwyrzykowski replied:
It creates a nullptr `WorkerClient`:

```
(lldb) bt
(lldb) bt
frame #0: WebCore`WebCore::ChromeClient::createWorkerClient(this=0x000000010bb0bd50, (null)=0x000000010bb1ee60) at ChromeClient.h:297:98 [opt]
frame #1: WebCore`WebCore::SharedWorkerThreadProxy::SharedWorkerThreadProxy(this=0x000000010bb278e0, page=<unavailable>, sharedWorkerIdentifier=WebCore::SharedWorkerIdentifier @ 0x000000016d6c9a40, clientOrigin=<unavailable>, workerFetchResult=<unavailable>, workerOptions=<unavailable>, initializationData=<unavailable>, cacheStorageProvider=<unavailable>) at SharedWorkerThreadProxy.cpp:116:46 [opt]
frame #2: WebKit`WebKit::WebSharedWorkerContextManagerConnection::launchSharedWorker(WebCore::ClientOrigin&&, WTF::ObjectIdentifierGeneric<WebCore::SharedWorkerIdentifierType, WTF::ObjectIdentifierMainThreadAccessTraits>, WebCore::WorkerOptions&&, WebCore::WorkerFetchResult&&, WebCore::WorkerInitializationData&&) [inlined] WTF::Ref<WebCore::SharedWorkerThreadProxy, WTF::RawPtrTraits<WebCore::SharedWorkerThreadProxy>, WTF::DefaultRefDerefTraits<WebCore::SharedWorkerThreadProxy>> WebCore::SharedWorkerThreadProxy::create<WTF::Ref<WebCore::Page, WTF::RawPtrTraits<WebCore::Page>, WTF::DefaultRefDerefTraits<WebCore::Page>>, WTF::ObjectIdentifierGeneric<WebCore::SharedWorkerIdentifierType, WTF::ObjectIdentifierMainThreadAccessTraits>&, WebCore::ClientOrigin&, WebCore::WorkerFetchResult, WebCore::WorkerOptions, WebCore::WorkerInitializationData, WebKit::WebCacheStorageProvider&>(args=0x000000016d6c9e28, args=<unavailable>, args=0x000000016d6ca150, args=0x000000016d6ca1a8, args=0x000000016d6ca198, args=0x000000016d6ca270, args=0x000000010b01c560) at SharedWorkerThreadProxy.h:53:114 [opt]
frame #3: WebKit::WebSharedWorkerContextManagerConnection::launchSharedWorker(this=0x000000010b8d8fe0, origin=0x000000016d6ca150, sharedWorkerIdentifier=WebCore::SharedWorkerIdentifier @ x24, workerOptions=0x000000016d6ca198, workerFetchResult=0x000000016d6ca1a8, initializationData=0x000000016d6ca270) at WebSharedWorkerContextManagerConnection.cpp:149:36 [opt]
```

ChromeClient::createWorkerClient is implemented as:
```
virtual std::unique_ptr<WorkerClient> createWorkerClient(SerialFunctionDispatcher&) { return nullptr; }
```

mwyrzykowski replied:
This works for dedicated workers because dedicated workers use a `WebChromeClient` but shared and service workers use an `EmptyChromeClient` it appears. I am unfamiliar with this code, just going off the debugger.

mattwoodrow replied:
Ah, I see!

I guess this really exposes an existing problem where the `createWorkerClient` function is tied into `ChromeClient`, where it probably should be a standalone `WorkerClientProvider`.

`WebSharedWorkerContextManagerConnection::launchSharedWorker` creates a page with an `EmptyChromeClient`, but does override a bunch of the other client/providers to be valid `WebKit` instances.

I think ideally we'd be able to do the same here, and have `pageConfiguration.workerClientProvider = makeUniqueRef<WorkerClientProvider>();`, and then the construction of the `WorkerClient` itself would work.

mwyrzykowski replied:
Oh nice! I’ll try that instead

mwyrzykowski replied:
@mattwoodrow I looked into implementing this and it seemed like WebWorkerClient didn't need the WebPage, just a WebCore::Page, so changing WebPage -> WebCore::Page allowed for Shared and Service workers to trivially construct a WebWorkerClient

Change is far simpler now and the conditional branch you mentioned above indeed just works now for shared and service workers :)

mwyrzykowski committed:
[WebGPU] WebGPU spec updated to support SharedWorkers and ServiceWorkers from only DedicatedWorkers
https://bugs.webkit.org/show_bug.cgi?id=273195
<radar://126990546>

Reviewed by Tadeu Zagallo.

Add support for SharedWorkers and ServiceWorkers as the spec
requires it now.

* Source/WebCore/Modules/WebGPU/GPU.idl:
* Source/WebCore/Modules/WebGPU/GPUAdapter.idl:
* Source/WebCore/Modules/WebGPU/GPUAdapterInfo.idl:
* Source/WebCore/Modules/WebGPU/GPUBindGroup.idl:
* Source/WebCore/Modules/WebGPU/GPUBindGroupLayout.idl:
* Source/WebCore/Modules/WebGPU/GPUBuffer.idl:
* Source/WebCore/Modules/WebGPU/GPUBufferUsage.idl:
* Source/WebCore/Modules/WebGPU/GPUColorWrite.idl:
* Source/WebCore/Modules/WebGPU/GPUCommandBuffer.idl:
* Source/WebCore/Modules/WebGPU/GPUCommandEncoder.idl:
* Source/WebCore/Modules/WebGPU/GPUCompilationInfo.idl:
* Source/WebCore/Modules/WebGPU/GPUCompilationMessage.idl:
* Source/WebCore/Modules/WebGPU/GPUComputePassEncoder.idl:
* Source/WebCore/Modules/WebGPU/GPUComputePipeline.idl:
* Source/WebCore/Modules/WebGPU/GPUDeviceLostInfo.idl:
* Source/WebCore/Modules/WebGPU/GPUExternalTexture.idl:
* Source/WebCore/Modules/WebGPU/GPUInternalError.idl:
* Source/WebCore/Modules/WebGPU/GPUMapMode.idl:
* Source/WebCore/Modules/WebGPU/GPUOutOfMemoryError.idl:
* Source/WebCore/Modules/WebGPU/GPUPipelineLayout.idl:
* Source/WebCore/Modules/WebGPU/GPUQuerySet.idl:
* Source/WebCore/Modules/WebGPU/GPUQueue.idl:
* Source/WebCore/Modules/WebGPU/GPURenderBundle.idl:
* Source/WebCore/Modules/WebGPU/GPURenderBundleEncoder.idl:
* Source/WebCore/Modules/WebGPU/GPURenderPassEncoder.idl:
* Source/WebCore/Modules/WebGPU/GPURenderPipeline.idl:
* Source/WebCore/Modules/WebGPU/GPUSampler.idl:
* Source/WebCore/Modules/WebGPU/GPUShaderModule.idl:
* Source/WebCore/Modules/WebGPU/GPUShaderStage.idl:
* Source/WebCore/Modules/WebGPU/GPUSupportedFeatures.idl:
* Source/WebCore/Modules/WebGPU/GPUSupportedLimits.idl:
* Source/WebCore/Modules/WebGPU/GPUTexture.idl:
* Source/WebCore/Modules/WebGPU/GPUTextureUsage.idl:
* Source/WebCore/Modules/WebGPU/GPUTextureView.idl:
* Source/WebCore/Modules/WebGPU/GPUUncapturedErrorEvent.idl:
* Source/WebCore/Modules/WebGPU/GPUValidationError.idl:
* Source/WebCore/Modules/WebGPU/WGSLLanguageFeatures.idl:
* Source/WebCore/html/canvas/GPUCanvasContext.idl:
Update idl files

* Source/WebCore/workers/WorkerGlobalScope.h:
* Source/WebCore/workers/WorkerLoaderProxy.h:
* Source/WebCore/workers/WorkerMessagingProxy.h:
* Source/WebCore/workers/WorkerThread.h:
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.cpp:
(WebCore::ServiceWorkerGlobalScope::create):
(WebCore::ServiceWorkerGlobalScope::ServiceWorkerGlobalScope):
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.h:
* Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
(WebCore::ServiceWorkerThread::createWorkerGlobalScope):
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.h:
* Source/WebCore/workers/shared/context/SharedWorkerThreadProxy.h:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::installServiceWorker):
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.h:
* Source/WebKit/WebProcess/Storage/WebSharedWorkerContextManagerConnection.cpp:
(WebKit::WebSharedWorkerContextManagerConnection::launchSharedWorker):
* Source/WebKit/WebProcess/Storage/WebSharedWorkerContextManagerConnection.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createWorkerClient):
* Source/WebKit/WebProcess/WebCoreSupport/WebWorkerClient.cpp:
(WebKit::WebWorkerClient::create):
* Source/WebKit/WebProcess/WebCoreSupport/WebWorkerClient.h:

Canonical link: https://commits.webkit.org/279984@main
